File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    let left = 1;
    let right = Math.max(...times) * n;

    let answer = right;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        let people = 0;

        for (const time of times) {
            people += Math.floor(mid / time);
            if (people >= n) break;
        }

        if (people >= n) {
            answer = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long left = 1;
        long right = (long) Arrays.stream(times).max().getAsInt() * n;

        long answer = right;

        while (left <= right) {
            long mid = (left + right) / 2;

            long people = 0;

            for (int time : times) {
                people += mid / time;
                if (people >= n) break; // 오버플로우 방지
            }

            if (people >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return answer;
    }
}